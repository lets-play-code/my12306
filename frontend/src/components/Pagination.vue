<template>
  <nav class="flex justify-center items-center my-4">
    <button
      @click="prevPage"
      :disabled="pageNum === 1 || totalPages <= 1"
      :class="[
        pageNum === 1 || totalPages <= 1 ? 'opacity-50 cursor-not-allowed' : '',
        'border border-gray-500 text-gray-500 rounded-md px-4 py-2 m-2 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline',
      ]"
    >
      Previous
    </button>
    <div v-if="totalPages > 1" class="flex">
      <button
        v-for="page in paginationNumbers"
        :key="page"
        @click="goToPage(page)"
        :disabled="pageNum === page"
        :class="[
          pageNum === page ? 'bg-gray-500 text-white' : 'text-gray-500',
          'border border-gray-500 rounded-md px-3 py-2 m-1 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline',
        ]"
      >
        {{ page === -1 ? "..." : page }}
      </button>
    </div>
    <button
      @click="nextPage"
      :disabled="pageNum === totalPages || totalPages <= 1"
      :class="[
        pageNum === totalPages || totalPages <= 1 ? 'opacity-50 cursor-not-allowed' : '',
        'border border-gray-500 text-gray-500 rounded-md px-4 py-2 m-2 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline',
      ]"
    >
      Next
    </button>
    <div class="flex items-center space-x-2">
      <span class="text-gray-500">Page {{ pageNum }} of {{ totalPages }}</span>
      <input data-testid="jump-page-input"
        v-model="jumpPageInput"
        @keydown.enter="jumpToPage"
        type="number"
        class="border border-gray-500 text-gray-500 rounded-md px-2 py-2 text-center"
        :max="totalPages"
        :min="1"
      />
      <button
        @click="jumpToPage"
        class="border border-gray-500 text-gray-500 rounded-md px-4 py-2 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline"
      >
        Go
      </button>
    </div>
  </nav>
</template>

<script setup lang="ts">
import { computed, ref, defineProps, defineEmits } from "vue";

const props = defineProps({
  pageNum: {
    type: Number,
    required: true
  },
  total: {
    type: Number,
    required: true
  },
  pageSize: {
    type: Number,
    default: 12
  }
});

const emit = defineEmits(["updatePage"]);

const totalPages = computed(() => Math.ceil(props.total / props.pageSize));

const prevPage = () => {
  if (props.pageNum > 1) {
    emit("updatePage", props.pageNum - 1);
  }
};

const nextPage = () => {
  if (props.pageNum < totalPages.value) {
    emit("updatePage", props.pageNum + 1);
  }
};

const goToPage = (page: number) => {
  if (page !== -1 && page !== props.pageNum) {
    emit("updatePage", page);
  }
};

const paginationNumbers = computed(() => {
  const visiblePages = 5;
  const pages = [];

  if (totalPages.value <= visiblePages) {
    for (let i = 1; i <= totalPages.value; i++) {
      pages.push(i);
    }
  } else {
    const halfVisiblePages = Math.floor(visiblePages / 2);
    if (props.pageNum - halfVisiblePages <= 1) {
      for (let i = 1; i <= visiblePages; i++) {
        pages.push(i);
      }
      pages.push(-1);
      pages.push(totalPages.value);
    } else if (props.pageNum + halfVisiblePages >= totalPages.value) {
      pages.push(1);
      pages.push(-1);
      for (let i = totalPages.value - visiblePages + 1; i <= totalPages.value; i++) {
        pages.push(i);
      }
    } else {
      pages.push(1);
      pages.push(-1);
      for (let i = props.pageNum - halfVisiblePages + 1; i <= props.pageNum + halfVisiblePages - 1; i++) {
        pages.push(i);
      }
      pages.push(-1);
      pages.push(totalPages.value);
    }
  }

  return pages;
});

const jumpPageInput = ref("");
const jumpToPage = () => {
  const pageNumber = parseInt(jumpPageInput.value);
  if (pageNumber >= 1 && pageNumber <= totalPages.value && pageNumber !== props.pageNum) {
    emit("updatePage", pageNumber);
  }
  jumpPageInput.value = "";
};
</script>
