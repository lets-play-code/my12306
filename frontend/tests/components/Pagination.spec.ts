import { render } from "./test-utils";
import { fireEvent } from "@testing-library/vue";
import { describe, it, expect } from "vitest";
import Pagination from "../../src/components/Pagination.vue";

describe("Pagination", () => {

  it("renders correctly", async () => {
    const { container, unmount } = render(Pagination, {
      props: {
        pageNum: 1,
        total: 20,
        pageSize: 5
      }
    });

    expect(container).toMatchSnapshot();
    unmount();
  });

  it("disables previous button on first page", async () => {
    const { getByText, emitted } = render(Pagination, {
      props: {
        pageNum: 1,
        total: 20,
        pageSize: 5
      }
    });

    const previousButton = getByText("Previous");
    expect(previousButton).toHaveProperty("disabled");
    await fireEvent.click(previousButton);
    expect(emitted().updatePage).toBeFalsy();
  });

  it("disables next button on last page", async () => {
    const { getByText, emitted } = render(Pagination, {
      props: {
        pageNum: 4,
        total: 20,
        pageSize: 5
      }
    });

    const nextButton = getByText("Next");
    expect(nextButton).toHaveProperty("disabled");
    await fireEvent.click(nextButton);
    expect(emitted().updatePage).toBeFalsy();
  });

  it("updates page when next button is clicked", async () => {
    const { getByText, emitted } = render(Pagination, {
      props: {
        pageNum: 1,
        total: 20,
        pageSize: 5
      }
    });

    const nextButton = getByText("Next");
    await fireEvent.click(nextButton);

    expect(emitted().updatePage).toBeTruthy();
    expect(emitted().updatePage[0]).toEqual([2]);
  });

  it("updates page when previous button is clicked", async () => {
    const { getByText, emitted } = render(Pagination, {
      props: {
        pageNum: 2,
        total: 20,
        pageSize: 5
      }
    });

    const previousButton = getByText("Previous");
    await fireEvent.click(previousButton);

    expect(emitted().updatePage).toBeTruthy();
    expect(emitted().updatePage[0]).toEqual([1]);
  });

  it("updates page when specific page is clicked", async () => {
    const { getByRole, emitted } = render(Pagination, {
      props: {
        pageNum: 1,
        total: 20,
        pageSize: 5
      }
    });

    const pageButton = getByRole("button", { exact: false, name: /^2$/i });
    console.log(pageButton.outerHTML);
    await fireEvent.click(pageButton);

    expect(emitted().updatePage).toBeTruthy();
    expect(emitted().updatePage[0]).toEqual([2]);
  });

  it("updates page when jump to specified page by input page number", async () => {
    const { getByRole, getByTestId, emitted } = render(Pagination, {
      props: {
        pageNum: 1,
        total: 20,
        pageSize: 5
      }
    });

    const input = getByTestId("jump-page-input");
    const button = getByRole("button", { name: "Go" });

    await fireEvent.update(input, "2");
    await fireEvent.click(button);

    expect(emitted().updatePage).toBeTruthy();
    expect(emitted().updatePage[0]).toEqual([2]);
  });
});
