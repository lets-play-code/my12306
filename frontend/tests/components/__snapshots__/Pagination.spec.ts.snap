// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > renders correctly 1`] = `
<div>
  <nav
    class="flex justify-center items-center my-4"
  >
    <button
      class="opacity-50 cursor-not-allowed border border-gray-500 text-gray-500 rounded-md px-4 py-2 m-2 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline"
      disabled=""
    >
       Previous 
    </button>
    <div
      class="flex"
    >
      
      <button
        class="bg-gray-500 text-white border border-gray-500 rounded-md px-3 py-2 m-1 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline"
        disabled=""
      >
        1
      </button>
      <button
        class="text-gray-500 border border-gray-500 rounded-md px-3 py-2 m-1 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline"
      >
        2
      </button>
      <button
        class="text-gray-500 border border-gray-500 rounded-md px-3 py-2 m-1 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline"
      >
        3
      </button>
      <button
        class="text-gray-500 border border-gray-500 rounded-md px-3 py-2 m-1 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline"
      >
        4
      </button>
      
    </div>
    <button
      class="border border-gray-500 text-gray-500 rounded-md px-4 py-2 m-2 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline"
    >
       Next 
    </button>
    <div
      class="flex items-center space-x-2"
    >
      <span
        class="text-gray-500"
      >
        Page 1 of 4
      </span>
      <input
        class="border border-gray-500 text-gray-500 rounded-md px-2 py-2 text-center"
        data-testid="jump-page-input"
        max="4"
        min="1"
        type="number"
      />
      <button
        class="border border-gray-500 text-gray-500 rounded-md px-4 py-2 transition duration-500 ease select-none hover:text-white hover:bg-gray-500 focus:outline-none focus:shadow-outline"
      >
         Go 
      </button>
    </div>
  </nav>
</div>
`;
